### PR TITLE
server/plugins/azure: use the default credentials API

### PR DESCRIPTION
**Pull request by shashankram**
Uses the NewDefaultAzureCredential API to fetch
client credentials. This API wraps different
mechanisms to obtain credentials using a chained
token credential mechanism. By doing so, the Azure plugins are able to obtain a token using any of the supported mechanisms: env vars, MSI, workload identity, without needing separate config input for each.

This is a part of #4485 to enable obtaining API tokens using Azure workload identity.

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

### CODE REVIEW

**Comment by shashankram** [comment]
@MarcosDY could you review this soon? This is blocking us from using SPIRE in Azure.

File: pkg/server/plugin/keymanager/azurekeyvault/azure_key_vault.go [hunk 1]
```diff
@@ -171,17 +171,20 @@ func (p *Plugin) Configure(ctx context.Context, req *configv1.ConfigureRequest)
...
 			return nil, status.Errorf(codes.Internal, "failed to create Key Vault client with client credentials: %v", err)
 		}
+
 	case config.UseMSI:
-		cred, err := p.hooks.msiCredential()
+		p.log.Warn("use_msi is deprecated and will be removed in a future release")
+		fallthrough // use default credential which attempts to fetch credentials using MSI
+
+	default:
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
```
Comment (MarcosDY) [inline pkg/server/plugin/keymanager/azurekeyvault/azure_key_vault.go:180]: can you add a test case for this? 
and may you remove the hook? or replace it with a new one but to replace direct use of azidentityNewDefaultAzureCredential

File: pkg/server/plugin/nodeattestor/azuremsi/msi.go [hunk 2]
```diff
@@ -203,11 +200,9 @@ func (p *MSIAttestorPlugin) Attest(stream nodeattestorv1.NodeAttestor_AttestServ
 	}
 
 	var selectorValues []string
-	if tenant.client != nil {
```
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/azuremsi/msi.go:206]: is it not longer possible to have a nil client?

File: pkg/server/plugin/nodeattestor/azuremsi/msi.go [hunk 3]
```diff
@@ -277,26 +272,34 @@ func (p *MSIAttestorPlugin) Configure(_ context.Context, req *configv1.Configure
...
+			p.log.Warn("use_msi is deprecated and will be removed in a future release")
+			fallthrough // use default credential which attempts to fetch credentials using MSI
+
+		default:
 			instanceMetadata, err := p.hooks.fetchInstanceMetadata(http.DefaultClient)
 			if err != nil {
 				return nil, err
 			}
-			cred, err := p.hooks.msiCredential()
+			cred, err := azidentity.NewDefaultAzureCredential(
```
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/azuremsi/msi.go:285]: can you add a test case for this failing?

File: pkg/server/plugin/nodeattestor/azuremsi/msi.go [hunk 4]
```diff
@@ -277,26 +272,34 @@ func (p *MSIAttestorPlugin) Configure(_ context.Context, req *configv1.Configure
...
 			}
 			client, err = p.hooks.newClient(instanceMetadata.Compute.SubscriptionID, cred)
 			if err != nil {
-				return nil, status.Errorf(codes.Internal, "unable to create client with MSI credential: %v", err)
+				return nil, status.Errorf(codes.Internal, "unable to create client with default credential: %v", err)
 			}
 		}
 
 		// If credentials are not configured then selectors won't be gathered.
 		// TODO: make this an error condition in a future release
```
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/azuremsi/msi.go:300]: can you remove todo?

File: pkg/server/plugin/nodeattestor/azuremsi/msi.go (hunk 2 above)
Comment (shashankram) [reply pkg/server/plugin/nodeattestor/azuremsi/msi.go:206]: If the plugin's `Configure()` fails, is it even possible to call `Attest()` on it? With this change, a nil client will result in `Configure()` producing an error.

File: pkg/server/plugin/nodeattestor/azuremsi/msi.go (hunk 3 above)
Comment (shashankram) [reply pkg/server/plugin/nodeattestor/azuremsi/msi.go:285]: With credential chaining happening in the SDK, errors are ignored. We would need to wrap this call behind a custom hook to test this scenario

File: pkg/server/plugin/nodeattestor/azuremsi/msi.go (hunk 4 above)
Comment (shashankram) [reply pkg/server/plugin/nodeattestor/azuremsi/msi.go]: done

File: pkg/server/plugin/nodeattestor/azuremsi/msi.go (hunk 2 above)
Comment (MarcosDY) [reply pkg/server/plugin/nodeattestor/azuremsi/msi.go:206]: you are right client is required now, so removing the if looks fine

**Comment by MarcosDY** [comment]
@shashankram can you resolve DCO?

**Comment by shashankram** [comment]
> @shashankram can you resolve DCO?

yes will address your comments and fix DCO

**Comment by shashankram** [comment]
@MarcosDY Had to force-push to fix DCO, I addressed the comments

File: pkg/server/plugin/keymanager/azurekeyvault/azure_key_vault.go (hunk 1 above)
Comment (shashankram) [reply pkg/server/plugin/keymanager/azurekeyvault/azure_key_vault.go]: I have modified the existing test. Otherwise, it's just an API wrapper from before.

**Review by MarcosDY — DISMISSED** [review]

**Comment by MarcosDY** [comment]
@shashankram code looks good, 
are you planning in update plugin documentation in a separated PR? (we need at least notify that use_msi is deprecated)

**Comment by shashankram** [comment]
> @shashankram code looks good, are you planning in update plugin documentation in a separated PR? (we need at least notify that use_msi is deprecated)

I'll push an update to this PR with the doc changes, thanks for the reminder!

**Comment by shashankram** [comment]
@MarcosDY I updated the docs, let me know if it makes sense to you

**Review by MarcosDY — APPROVED** [review]

**Comment by shashankram** [comment]
There seems to be a few unrelated test flakes. How do I rerun these?

**Pull request merged**